### PR TITLE
Sorting Improvments

### PR DESCRIPTION
**Pull request by jamiewhths**
- Refactor HeapsortColumn to include "ascending" parameter and remove redundant methods
- Added isLessThan() methods to encapsulate handling of missing values for each column type and comparison of values for asc and desc sorts
- Implement IntColumnView8/IntColumnView16 column sort methods
- Implement BitSet/BitSetWithMissing and SparseNumberColumn sort methods
- Improve String collation by lower-casing before comparison
- Improve test cases with assertions and expand to cover each column sort method
- Fix broken "Clear Sort" menu option
- Fix sort icon overflow issue

### CODE REVIEW

File: model/build.gradle [hunk 1]
```diff
@@ -14,6 +14,11 @@ dependencies {
     compile libraries.jackson.core
 
     compile libraries.swagger.annotations
+
+    // Primitive Collection Library for speed and lower-memory
+    // footprint
+    // UPGRADE to latest version when we can use Java 8
+    compile 'it.unimi.dsi:fastutil:7.2.1'
```
Comment (akbertram) [inline model/build.gradle]: We are now building with Java 8.

File: model/src/main/java/org/activityinfo/model/query/BitSetColumnView.java [hunk 2]
```diff
@@ -98,8 +100,22 @@ public ColumnView select(int[] selectedRows) {
...
     @Override
     public int[] order(int[] sortVector, SortModel.Dir direction, int[] range) {
-        // TODO: BitSet Sorting
-        // Do not sort on column
+        switch(direction) {
+            case ASC:
+                if (range == null || range.length == numRows) {
+                    HeapsortColumn.heapsortBitSet(bitSet, sortVector, numRows, true);
+                } else {
+                    HeapsortColumn.heapsortBitSet(bitSet, sortVector, range.length, range, true);
```
Comment (akbertram) [inline model/src/main/java/org/activityinfo/model/query/BitSetColumnView.java]: Can we collapse these two branches and just pass the parameter (direction == Direction.ASC) ?

File: model/src/main/java/org/activityinfo/model/util/HeapsortColumn.java [hunk 3]
```diff
@@ -19,186 +19,208 @@
...
-        } else if (enum1 == -1) {
-            return -1;
-        } else if (enum2 == -1) {
-            return 1;
+    private static boolean isLessThan(short sx, short sy, boolean ascending) {
+        // Treat as unsigned - promote to int with a 16-bit mask
+        int x = sx & 0xFFFF;
+        int y = sy & 0xFFFF;
+        if(ascending) {
+            if(x == 0 && y != 0) {
```
Comment (akbertram) [inline model/src/main/java/org/activityinfo/model/util/HeapsortColumn.java]: This probably would benefit from some commentary... we will both forget what this line does in 6 months. One way to assign the expressions to variables to give them names...

```
// missing values encoded as zeros
boolean xMissing = (x == 0);
boolean yMissing = (y == 0);
```

File: model/build.gradle (hunk 1 above)
Comment (jamiewhths) [reply model/build.gradle]: This was a copy-paste from the other module. Should we upgrade both to use the latest?

File: model/src/main/java/org/activityinfo/model/query/BitSetColumnView.java (hunk 2 above)
Comment (jamiewhths) [reply model/src/main/java/org/activityinfo/model/query/BitSetColumnView.java]: Yes I actually did that with the latest commit, but I should update the rest as well.

File: model/src/main/java/org/activityinfo/model/util/HeapsortColumn.java (hunk 3 above)
Comment (jamiewhths) [reply model/src/main/java/org/activityinfo/model/util/HeapsortColumn.java]: Will do

File: model/build.gradle (hunk 1 above)
Comment (akbertram) [reply model/build.gradle]: Yes that would probably be a good idea. You can define the dependency in libraries.gradle to keep the dependencies in sync.

**Pull request merged**